opm/autodiff/ExtendedBlackoilModel_impl.hpp

```diff
@@ -0,0 +1,703 @@
+        std::array<double,MaxNumPhases> well_flux_residual    = {{0., 0., 0.}};
+        std::size_t cols = MaxNumPhases+1; // needed to pass the correct type to Eigen
+        Eigen::Array<V::Scalar, Eigen::Dynamic, MaxNumPhases+1> B(nc, cols);
+        Eigen::Array<V::Scalar, Eigen::Dynamic, MaxNumPhases+1> R(nc, cols);
+        Eigen::Array<V::Scalar, Eigen::Dynamic, MaxNumPhases+1> tempV(nc, cols);
+        std::vector<double> maxNormWell(MaxNumPhases);
+
+        for ( int idx=0; idx<MaxNumPhases; ++idx )
+        {
+            if (active_[idx]) {
+                const int pos    = pu.phase_pos[idx];
+                const ADB& tempB = rq_[pos].b;
+                B.col(idx)       = 1./tempB.value();
+                R.col(idx)       = residual_.material_balance_eq[idx].value();
+                tempV.col(idx)   = R.col(idx).abs()/pv;
+            }
+        }
+        if (has_solvent_) {
+            const ADB& tempB = rq_[solvent_pos_].b;
+            B.col(MaxNumPhases) = 1. / tempB.value();
+            R.col(MaxNumPhases) = residual_.material_balance_eq[solvent_pos_].value();
+            tempV.col(MaxNumPhases) = R.col(MaxNumPhases).abs()/pv;
+        }
+
+        const double pvSum = convergenceReduction(B, tempV, R, R_sum, maxCoeff, B_avg,
+                                                  maxNormWell, nc, nw);
+
+        bool converged_MB = true;
+        bool converged_CNV = true;
+        bool converged_Well = true;
+        // Finish computation
+        for ( int idx = 0; idx < (MaxNumPhases + 1) ; ++idx )
+        {
+            CNV[idx]                   = B_avg[idx] * dt * maxCoeff[idx];
+            mass_balance_residual[idx] = std::abs(B_avg[idx]*R_sum[idx]) * dt / pvSum;
+            converged_MB               = converged_MB && (mass_balance_residual[idx] < tol_mb);
+            converged_CNV              = converged_CNV && (CNV[idx] < tol_cnv);
+            if (idx != MaxNumPhases) { // No well flux residual for polymer.
+                well_flux_residual[idx]    = B_avg[idx] * maxNormWell[idx];
+                converged_Well = converged_Well && (well_flux_residual[idx] < tol_wells);
+            }
+        }
+
+        const double residualWell     = detail::infinityNormWell(residual_.well_eq,
+                                                                 linsolver_.parallelInformation());
+        converged_Well   = converged_Well && (residualWell < Opm::unit::barsa);
+        const bool   converged        = converged_MB && converged_CNV && converged_Well;
+
+        // if one of the residuals is NaN, throw exception, so that the solver can be restarted
+        if (std::isnan(mass_balance_residual[Water]) || mass_balance_residual[Water] > maxResidualAllowed() ||
+            std::isnan(mass_balance_residual[Oil])   || mass_balance_residual[Oil]   > maxResidualAllowed() ||
+            std::isnan(mass_balance_residual[Gas])   || mass_balance_residual[Gas]   > maxResidualAllowed() ||
```

Comment:
The solvent residual should be checked here, should it not? Also CNV a few lines down. Not sure about the well flux residuals though: there is no separate solvent well flux equation?